### DISCUSSION TITLE
Keep certain files open

### DISCUSSION THREAD

**Discussion by ElMoribond** [alefragnani/vscode-project-manager#614] [Q&A] [answered]
Your extension is awesome, thank you
I use it daily to develop and machines administration.

There's one thing I wish I could do.
Change project while keeping some files from the previous project open.

Sometimes a configuration file is better than a long speech.
Imagine that we have
```[
	{
		"name": "Host_0 etc",
		"rootPath": "vscode-remote://ssh-remote+Host_0/etc",
		"paths": [],
		"tags": [],
		"enabled": true,
		"toDoWhenLeaving": "keep_all_open"
	},
	{
		"name": "Host_0 srv",
		"rootPath": "vscode-remote://ssh-remote+Host_0/srv",
		"paths": [],
		"tags": [],
		"enabled": true,
		"toDoWhenLeaving": "keep_modified_open"
	},
	{
		"name": "Host_1 etc",
		"rootPath": "vscode-remote://ssh-remote+Host_1/etc",
		"paths": [],
		"tags": [],
		"enabled": true,
		"toDoWhenLeaving": "keep_all_open_only_same_host"
	},
	{
		"name": "Host_2 etc",
		"rootPath": "vscode-remote://ssh-remote+Host_2/etc",
		"paths": [],
		"tags": [],
		"enabled": true,
		"toDoWhenLeaving": "close_all"
	}
]
```

**Comment by alefragnani** [comment, marked as answer]
Hi @ElMoribond ,

Thank you for your kind words. I'm glad to know the extension is being helpful.

I guess I understand your idea, and indeed it can be useful. And I suppose it would only work/be available when you open the project _on the same window_, right?

But I see an issue/complication about how to achieve it, because _one VS Code instance doesn't know about others or previous windows_. It means that if you are working in **Host_0 etc** and switch to **Host_0 srv**, neither VS Code nor the extension knows that _you came from "Host_0 etc"_. Also, when opening **Host_0 srv**, VS Code itself will restore the session, and reopen the files you had open on previous session. Opening _other files_ could conflict with this restore feature.

With that in mind, I'm not sure such feature is doable. At least not without refining and further investigation.

Hope this helps

**Reply by ElMoribond** [reply]
Hello Alessandro,

You understood the idea well and I was a little afraid of your answer.

You are right, if there is no way to "manage/configure" the sessions it seems difficult to me.
Not to say impossible.

Your explanations are very clear. Thank you)

> I guess I understand your idea, and indeed it can be useful. And I suppose it would only work/be available when you open the project on the same window, right?


Not necessarily, because if I understood correctly, the process is not fundamentally different.
In any case there is a new VsCode session.


I have no knowledge of VsCode extensions and I may sound delusional to you:

When we change projects, the extension knows where we come from and where we want to go.
Via the value of toDoWhenLeaving it can therefore generate "a piece of code" which will determine the actions.

These actions are in fact exclusively file openings (possibly from another machine).
- only if they have not already been restored by VsCode
- if there are unsaved files we put them in "the piece of code"

Then, when launching the new VsCode session, we also send "the piece of code".
Which is picked up by the new instance of the extension.

In short, you have the skills to judge and if you tell me that such a scenario is not possible, well, it's just sad but not dramatic.
I would continue to use your extension as it saves me a lot of time.

Michael

**Reply by alefragnani** [reply]
Hi @ElMoribond , 

I'm happy to see you understand some of the limitations the VS Code API imposes.

On the other hand, I wouldn't say it is _impossible_, but it wouldn't have any connection with the VS Code API. Maybe, it could use a _state_ concept, where the first window _stores_ its state (the opened files) and whenever a new window is opened, it _loads_ the state, to check for "files to open". 

There is however, based on your samples, the possibility of issues with the remote files/folders you have. I'm not sure how one could work while trying to open from another (I don't do too much remote access lately), and I guess it would have some issues related to the base folder/location, but just testing to be sure.

With that in mind, I wonder if such feature isn't better suited for a _new extension_ 🤔
